File: components/platform/ts/src/js/host.js
// The seam between the JS runtime and the native side.
//
// Everything the runtime needs from Rust — creating views, applying modifiers,
// control flow, the environment — goes through the object installed by
// `installHost`. The bridge installs it right after evaluating the bundle and
// before any module is mounted; the contract it must satisfy is specified in
// HOST.md next to this file.
//
// The other exports are the bridge's materialization helpers: the Rust host
// table calls `isSignal` / `isAccessor` / `read` / `write` / `subscribe` /
// `toSignal` to turn JS values into real `Binding<T>` / `Computed<T>` objects.

import {
  bridgeEquals,
  comparatorOf,
  createEffect,
  createMemo,
  createRoot,
  createSignal,
  isAccessor,
  isMemo,
  isSignal,
  onCleanup,
  untrack,
} from "./signals.js";

export { isAccessor, isMemo, isSignal };

/**
 * @typedef {unknown} Handle
 * An opaque native view handle. The runtime never inspects it; only the host
 * interprets it. See HOST.md.
 */

/**
 * @typedef {object} Branch
 * One realized branch of a control-flow view, created under its own scope.
 * @property {Handle} handle - The native view to mount for this branch.
 * @property {() => void} dispose - Tears the branch down; the host calls it
 *   exactly once when the branch stops being presented.
 */

/**
 * @typedef {object} HostEnvironment
 * @property {unknown} theme - Reactive-or-plain theme value; see `toSignal`.
 * @property {unknown} locale - Reactive-or-plain locale value.
 */

/**
 * @typedef {object} Host
 * The native side of the runtime. Specified in full in HOST.md. `when`,
 * `each`, `content`, config values, and modifier values are reactive inputs:
 * `T | Signal<T> | (() => T) | { read, subscribe? }`.
 * @property {(component: string, config: object, children: unknown[]) => Handle} create
 * @property {(handle: Handle, name: string, value: unknown) => Handle} modify
 * @property {(content: unknown) => Handle} text
 * @property {(when: unknown, render: (item: () => unknown) => Branch, fallback?: () => Branch) => Handle} show
 * @property {(each: unknown, render: (item: unknown, index: () => number) => Branch, by?: (item: unknown) => unknown) => Handle} each
 * @property {(children: () => Branch, fallback?: () => Branch) => Handle} suspense
 * @property {() => HostEnvironment} environment
 * @property {(id: number, ...args: unknown[]) => unknown} invoke - Dispatches
 *   one Rust closure held in the bridge's registry. It travels with the table
 *   rather than being read off `globalThis.__waterui_host` at call time,
 *   because that global is writable and a bundle that reassigns an entry of it
 *   while it evaluates would otherwise be what every callback wrapper calls.
 * @property {ReadonlySet<string> | readonly string[]} modifiers - The
 *   catalog's modifier attribute names. The runtime splits config from
 *   modifier attributes against it and rejects spread-carried modifiers; the
 *   Rust host table provides it because the catalog is the single source of
 *   truth. A `Set` cannot cross the engine seam as a value, so the Rust table
 *   sends the names as an array and `installHost` builds the set once.
 */

let installedHost = null;

/**
 * Installs the native host table. Called once by the bridge, after the bundle
 * is evaluated and before any module is mounted; every component creation
 * flows through it.
 *
 * The runtime keeps the table with its `modifiers` entry normalized to a real
 * `Set`, which may be a copy of the object passed in — a host entry must
 * therefore be a plain function and must not depend on `this`.
 *
 * @param {import("./host.js").Host} host
 */
export function installHost(host) {
  if (installedHost !== null) {
    throw new Error("waterui: a host is already installed");
  }
  for (const name of [
    "create",
    "modify",
    "text",
    "show",
    "each",
    "suspense",
    "environment",
    "invoke",
  ]) {
    if (typeof host[name] !== "function") {
      throw new TypeError(`waterui host is missing the "${name}" entry — see HOST.md`);
    }
  }
  if (Array.isArray(host.modifiers)) {
    installedHost = { ...host, modifiers: new Set(host.modifiers) };
    return;
  }
  if (
    host.modifiers === null ||
    typeof host.modifiers !== "object" ||
    typeof host.modifiers.has !== "function"
  ) {
    throw new TypeError(
      'waterui host is missing the "modifiers" entry — the catalog\'s modifier names as an array or a ReadonlySet<string>, see HOST.md',
    );
  }
  installedHost = host;
}

/** Removes the installed host. The bridge calls this when unloading a bundle. */
export function uninstallHost() {
  installedHost = null;
}

/** The installed host, or a precise error when the bundle ran without one. */
export function getHost() {
  if (installedHost === null) {
    throw new Error(
      "waterui: no host installed — installHost() must run before a module is mounted",
    );
  }
  return installedHost;
}

// ---------------------------------------------------------------------------
// Materialization helpers — what the Rust bridge calls.
// ---------------------------------------------------------------------------

function trackedRead(source) {
  return typeof source === "function" ? source() : source.read();
}

/**
 * Reads `value` without recording a dependency. Constants pass through, so
 * `read` also answers "the current value of anything positionally dynamic".
 */
export function read(value) {
  if (typeof value === "function" || (value !== null && typeof value === "object" && typeof value.read === "function")) {
    return untrack(() => trackedRead(value));
  }
  return value;
}

/**
 * Writes `value` into a signal or a host-side writable reactive value, and
 * answers whether the value stood: `true` when reading the target back gives
 * exactly what was written, `false` when an effect changed it while the write
 * settled. Anything else — memos, plain accessors, constants — is read-only
 * and fails loudly rather than being silently dropped.
 *
 * The bridge needs that answer and cannot work it out for itself. A write
 * settles its effects synchronously, and one of them may clamp, round or
 * reject the value; the comparison that catches it turns on object identity,
 * and the same signal or callback crossing back out to Rust is a fresh
 * reference there. Here it is the same object, so the comparison is exact.
 *
 * A value equal to what the target already holds — by `bridgeEquals`, the
 * seam's structural equality — is not written at all, and stands by
 * definition. The bridge confirms every inbound change by writing back what
 * the native side settled on, and that confirmation must cost no propagation:
 * the payload is a fresh object every time it crosses, so a reference
 * comparison would announce a change to every subscriber for a value nobody
 * touched. Values that do differ are written, and the target's own comparator
 * decides whether that write is a change.
 *
 * `identity` turns that comparison into `Object.is`. The native side passes
 * it when the value it sent is a retained handle — a live object, a function,
 * an opaque value — rather than data: a handle crosses as itself, so two
 * handles that look alike are still two different things, and JavaScript
 * cannot tell one from a plain object on its own.
 *
 * @param {unknown} target
 * @param {unknown} value
 * @param {boolean} [identity] - Compare with `Object.is` instead of
 *   structurally, because `value` crossed as a handle.
 * @returns {boolean} Whether the target now holds exactly `value`.
 */
export function write(target, value, identity = false) {
  const alreadyHolds = (held) => (identity ? Object.is(held, value) : bridgeEquals(held, value));
  if (isSignal(target)) {
    if (alreadyHolds(read(target))) {
      return true;
    }
    // An updater, not the value: `set` calls a function argument with the old
    // value and stores the result, so a callback or a signal written straight
    // through it would be invoked instead of stored. An updater that returns
    // the value stores it verbatim, whatever its type.
    target.set(() => value);
    return comparatorOf(target)(read(target), value);
  }
  if (target !== null && typeof target === "object" && typeof target.write === "function") {
    if (alreadyHolds(read(target))) {
      return true;
    }
    target.write(value);
    return comparatorOf(target)(read(target), value);
  }
  throw new TypeError("write() expects a signal or a writable reactive value");
}

/**
 * Runs `callback(value)` on every settled change of `source`, without an
 * initial call. Sources are signals, memos, thunks (tracked), or host-side
 * `{ read, subscribe }` values. Returns a function that unsubscribes and
 * disposes the subscription.
 */
export function subscribe(source, callback) {
  if (
    source !== null &&
    typeof source === "object" &&
    typeof source.subscribe === "function"
  ) {
    return source.subscribe(callback);
  }
  if (typeof source !== "function") {
    throw new TypeError(
      "subscribe() expects a signal, an accessor, or a { read, subscribe } value",
    );
  }
  return createRoot((dispose) => {
    let first = true;
    createEffect(() => {
      const value = source();
      if (first) {
        first = false;
        return;
      }
      callback(value);
    });
    return dispose;
  });
}

/**
 * Lifts `T | Signal<T> | (() => T)` to an accessor. Anything callable passes
 * through untouched; a `{ read() }` host value is wrapped so the result is
 * always callable — use `toSignal` instead when the subscription matters.
 */
export function toAccessor(value) {
  if (typeof value === "function") {
    return value;
  }
  if (value !== null && typeof value === "object" && typeof value.read === "function") {
    return () => value.read();
  }
  return () => value;
}

/**
 * Materializes any reactive-or-constant input as a real `Signal<T>`. Signals
 * and memos pass through; host-side `{ read, subscribe }` values become a
 * push-fed signal whose subscription is released with the current owner;
 * thunks become memos; constants become constant signals.
 */
export function toSignal(source) {
  if (isSignal(source) || isMemo(source)) {
    return source;
  }
  if (
    source !== null &&
    typeof source === "object" &&
    typeof source.read === "function"
  ) {
    const signal = createSignal(source.read());
    if (typeof source.subscribe === "function") {
      // An updater, for the same reason `write` uses one: a host value
      // carrying a function must be stored, not called.
      onCleanup(source.subscribe((value) => signal.set(() => value)));
    }
    return signal;
  }
  if (typeof source === "function") {
    return createMemo(source);
  }
  return createSignal(source);
}

File: components/platform/ts/src/host.rs
//! The native side of the runtime: the host table, and how it is installed.
//!
//! [`HostTable`] is the Rust half of the `Host` interface in HOST.md. The
//! component bindings that implement it — resolving `"VStack"` against the
//! catalog, applying `"padding"`, reconciling `each` — are the host-table leaf
//! (water-rs/waterui#1042); what lives here is the seam: the host functions
//! the engine registers, the argument shapes they hold JavaScript to, and the
//! table object `installHost` receives.
//!
//! Two entries never reach the table. `environment()` is answered from the
//! `Environment` the runtime was constructed with, because the framework owns
//! those values, and `invoke` is the bridge's own: it dispatches every
//! registered Rust closure, from a prop callback to a signal subscription.

use std::rc::Rc;

use suiteki::Str;
use waterui_core::{AnyView, Error};
use waterui_ts_engine::{JsError, JsFunction, JsRuntime, JsValue};

use crate::bridge::{Bridge, WeakBridge};
use crate::environment::host_environment;
use crate::error::{TsError, kind_of};
use crate::view::ViewSlot;

/// The native views a TypeScript module's JSX turns into.
///
/// Every value a method is handed is exactly what HOST.md describes: `config`
/// entries and modifier values are reactive inputs, materialized with
/// [`Bridge::materialize_binding`] or [`Bridge::materialize_computed`] at the
/// moment they reach a native view; `children` are view slots, strings,
/// numbers or accessors; `render` callbacks return `{ handle, dispose }`
/// branches the implementation drives.
///
/// An error is thrown into JavaScript where the call was made, with its
/// message intact.
pub trait HostTable: 'static {
    /// The attribute names the component catalog declares as modifiers.
    ///
    /// Read once at install: the runtime keeps the set and classifies
    /// attributes locally, so this never costs a crossing per element.
    fn modifiers(&self) -> Vec<Str>;

    /// Creates a native view for a JSX element.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] for a component the catalog does not have, or a
    /// configuration value the component cannot take.
    fn create(
        &self,
        bridge: &Bridge,
        component: &str,
        config: &[(String, JsValue)],
        children: &[JsValue],
    ) -> Result<AnyView, Error>;

    /// Applies one modifier, in the order the attributes were written.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] for a modifier the catalog does not have, or a value
    /// it cannot take.
    fn modify(
        &self,
        bridge: &Bridge,
        view: AnyView,
        name: &str,
        value: &JsValue,
    ) -> Result<AnyView, Error>;

    /// Materializes a text leaf, localized like Rust's own `text(…)`.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] when the content is neither text nor a reactive value
    /// producing text.
    fn text(&self, bridge: &Bridge, content: &JsValue) -> Result<AnyView, Error>;

    /// `<Show>`: presents one branch or the other as `when` changes.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] when a branch cannot be rendered.
    fn show(
        &self,
        bridge: &Bridge,
        when: &JsValue,
        render: &JsFunction,
        fallback: Option<&JsFunction>,
    ) -> Result<AnyView, Error>;

    /// `<For>`: reconciles a list by identity.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] when the list or a branch cannot be rendered.
    fn each(
        &self,
        bridge: &Bridge,
        items: &JsValue,
        render: &JsFunction,
        by: Option<&JsFunction>,
    ) -> Result<AnyView, Error>;

    /// `<Suspense>`: presents the fallback while the children are pending.
    ///
    /// # Errors
    ///
    /// Returns [`Error`] when a branch cannot be rendered.
    fn suspense(
        &self,
        bridge: &Bridge,
        children: &JsFunction,
        fallback: Option<&JsFunction>,
    ) -> Result<AnyView, Error>;
}

/// The property of `__waterui_host` one host function is registered under,
/// and the expression that reads it back for the table object.
struct HostEntry {
    name: &'static str,
    source: &'static str,
}

const CREATE: HostEntry = HostEntry {
    name: "create",
    source: "globalThis.__waterui_host.create",
};
const MODIFY: HostEntry = HostEntry {
    name: "modify",
    source: "globalThis.__waterui_host.modify",
};
const TEXT: HostEntry = HostEntry {
    name: "text",
    source: "globalThis.__waterui_host.text",
};
const SHOW: HostEntry = HostEntry {
    name: "show",
    source: "globalThis.__waterui_host.show",
};
const EACH: HostEntry = HostEntry {
    name: "each",
    source: "globalThis.__waterui_host.each",
};
const SUSPENSE: HostEntry = HostEntry {
    name: "suspense",
    source: "globalThis.__waterui_host.suspense",
};
const ENVIRONMENT: HostEntry = HostEntry {
    name: "environment",
    source: "globalThis.__waterui_host.environment",
};

/// The entry every registered Rust closure is dispatched through.
const INVOKE: &str = "invoke";

/// The script name the host reads are attributed to in a stack trace.
const SOURCE_NAME: &str = "waterui:host";

/// Registers every host function on the engine.
///
/// Registration happens before the bundle is evaluated, so the functions exist
/// no matter when JavaScript first reaches for one.
///
/// # Errors
///
/// Returns [`JsError`] when the engine cannot install a function.
pub fn register(bridge: &Bridge, table: &Rc<dyn HostTable>) -> Result<(), JsError> {
    let engine = bridge.engine();

    engine.register(INVOKE, {
        let callbacks = Rc::clone(bridge.callbacks());
        move |args: &[JsValue]| {
            let id = args
                .first()
                .and_then(JsValue::as_u64)
                .and_then(|id| u32::try_from(id).ok())
                .ok_or_else(|| {
                    JsError::conversion(
                        "__waterui_host.invoke(id, …) takes the callback id makeCallback was \
                         given as its first argument",
                    )
                })?;
            callbacks.invoke(id, args.get(1..).unwrap_or_default())
        }
    })?;

    engine.register(
        CREATE.name,
        host_call(bridge, table, |table, bridge, args| {
            let component = string_argument(args, 0, "create(component, config, children)")?;
            let config = object_argument(args, 1, "create(component, config, children)")?;
            let children = array_argument(args, 2, "create(component, config, children)")?;
            table.create(bridge, component, config, children)
        }),
    )?;

    engine.register(
        MODIFY.name,
        host_call(bridge, table, |table, bridge, args| {
            let view = bridge.take_view(argument(args, 0, "modify(handle, name, value)")?)?;
            let name = string_argument(args, 1, "modify(handle, name, value)")?;
            let value = argument(args, 2, "modify(handle, name, value)")?;
            table.modify(bridge, view, name, value)
        }),
    )?;

    engine.register(
        TEXT.name,
        host_call(bridge, table, |table, bridge, args| {
            table.text(bridge, argument(args, 0, "text(content)")?)
        }),
    )?;

    engine.register(
        SHOW.name,
        host_call(bridge, table, |table, bridge, args| {
            let when = argument(args, 0, "show(when, render, fallback?)")?;
            let render = function_argument(args, 1, "show(when, render, fallback?)")?;
            let fallback = optional_function_argument(args, 2, "show(when, render, fallback?)")?;
            table.show(bridge, when, render, fallback)
        }),
    )?;

    engine.register(
        EACH.name,
        host_call(bridge, table, |table, bridge, args| {
            let items = argument(args, 0, "each(each, render, by?)")?;
            let render = function_argument(args, 1, "each(each, render, by?)")?;
            let by = optional_function_argument(args, 2, "each(each, render, by?)")?;
            table.each(bridge, items, render, by)
        }),
    )?;

    engine.register(
        SUSPENSE.name,
        host_call(bridge, table, |table, bridge, args| {
            let children = function_argument(args, 0, "suspense(children, fallback?)")?;
            let fallback = optional_function_argument(args, 1, "suspense(children, fallback?)")?;
            table.suspense(bridge, children, fallback)
        }),
    )?;

    engine.register(ENVIRONMENT.name, {
        let weak = bridge.downgrade();
        move |_args: &[JsValue]| host_environment(&upgrade(&weak)?)
    })?;

    Ok(())
}

/// Installs the table object on the JavaScript side.
///
/// Called once, after the bundle is evaluated and before anything is mounted:
/// the installer itself lives in the bundle, so it cannot run earlier.
///
/// # Errors
///
/// Returns [`TsError`] when a host function cannot be read back, or when
/// `installHost` rejects the table.
pub fn install(bridge: &Bridge, table: &Rc<dyn HostTable>) -> Result<(), TsError> {
    let entries = [CREATE, MODIFY, TEXT, SHOW, EACH, SUSPENSE, ENVIRONMENT];
    let mut object = Vec::with_capacity(entries.len() + 1);
    for entry in entries {
        match bridge.engine().eval(entry.source, SOURCE_NAME)? {
            JsValue::Function(function) => {
                object.push((entry.name.to_owned(), JsValue::Function(function)));
            }
            other => {
                return Err(TsError::RuntimeEntry {
                    name: entry.name,
                    found: kind_of(&other),
                });
            }
        }
    }
    // A `Set` is not a plain object and cannot cross the seam, so the names
    // travel as an array and `installHost` builds the set once.
    object.push((
        String::from("modifiers"),
        JsValue::Array(
            table
                .modifiers()
                .into_iter()
                .map(|name| JsValue::String(name.as_str().to_owned()))
                .collect(),
        ),
    ));

    let runtime = bridge.runtime()?;
    bridge.call(runtime.install_host(), &[JsValue::Object(object)])?;
    Ok(())
}

/// Wraps one table method as a host function: upgrade, call, wrap the view.
fn host_call(
    bridge: &Bridge,
    table: &Rc<dyn HostTable>,
    call: impl Fn(&dyn HostTable, &Bridge, &[JsValue]) -> Result<AnyView, Error> + 'static,
) -> impl Fn(&[JsValue]) -> Result<JsValue, JsError> + 'static {
    let weak = bridge.downgrade();
    let table = Rc::clone(table);
    move |args: &[JsValue]| {
        let bridge = upgrade(&weak)?;
        let view = call(table.as_ref(), &bridge, args)
            .map_err(|error| JsError::new("Error", format!("{error:#}")))?;
        Ok(ViewSlot::new(view).to_js_value())
    }
}

/// The bridge, or the error JavaScript sees once the runtime is gone.
fn upgrade(weak: &WeakBridge) -> Result<Bridge, JsError> {
    weak.upgrade().ok_or_else(|| {
        JsError::new(
            "Error",
            "the TypeScript runtime was dropped while JavaScript was calling into it",
        )
    })
}

/// One argument, or an error naming the call that wanted it.
fn argument<'a>(args: &'a [JsValue], index: usize, call: &str) -> Result<&'a JsValue, JsError> {
    args.get(index).ok_or_else(|| {
        JsError::conversion(format!(
            "{call} was called with fewer than {} arguments",
            index + 1
        ))
    })
}

fn string_argument<'a>(args: &'a [JsValue], index: usize, call: &str) -> Result<&'a str, JsError> {
    let value = argument(args, index, call)?;
    value.as_str().ok_or_else(|| {
        JsError::conversion(format!(
            "{call}: argument {index} is {}, not a string",
            kind_of(value)
        ))
    })
}

fn object_argument<'a>(
    args: &'a [JsValue],
    index: usize,
    call: &str,
) -> Result<&'a [(String, JsValue)], JsError> {
    let value = argument(args, index, call)?;
    value.as_object().ok_or_else(|| {
        JsError::conversion(format!(
            "{call}: argument {index} is {}, not an object",
            kind_of(value)
        ))
    })
}

fn array_argument<'a>(
    args: &'a [JsValue],
    index: usize,
    call: &str,
) -> Result<&'a [JsValue], JsError> {
    let value = argument(args, index, call)?;
    value.as_array().ok_or_else(|| {
        JsError::conversion(format!(
            "{call}: argument {index} is {}, not an array",
            kind_of(value)
        ))
    })
}

fn function_argument<'a>(
    args: &'a [JsValue],
    index: usize,
    call: &str,
) -> Result<&'a JsFunction, JsError> {
    match argument(args, index, call)? {
        JsValue::Function(function) => Ok(function),
        other => Err(JsError::conversion(format!(
            "{call}: argument {index} is {}, not a function",
            kind_of(other)
        ))),
    }
}

/// An optional trailing function argument: absent, `undefined` and `null` all
/// mean the caller did not pass one.
fn optional_function_argument<'a>(
    args: &'a [JsValue],
    index: usize,
    call: &str,
) -> Result<Option<&'a JsFunction>, JsError> {
    match args.get(index) {
        None | Some(JsValue::Undefined | JsValue::Null) => Ok(None),
        Some(JsValue::Function(function)) => Ok(Some(function)),
        Some(other) => Err(JsError::conversion(format!(
            "{call}: argument {index} is {}, not a function",
            kind_of(other)
        ))),
    }
}
